llvm/lib/DebugInfo/DWARF/DWARFContext.cpp

```diff
@@ -89,9 +89,12 @@ void fixupIndexV4(DWARFContext &C, DWARFUnitIndex &Index) {
     DWARFDataExtractor Data(DObj, S, C.isLittleEndian(), 0);
     while (Data.isValidOffset(Offset)) {
       DWARFUnitHeader Header;
-      if (!Header.extract(C, Data, &Offset, DWARFSectionKind::DW_SECT_INFO)) {
+      if (Error ExtractionErr = Header.extract(
+              C, Data, &Offset, DWARFSectionKind::DW_SECT_INFO)) {
         logAllUnhandledErrors(
-            createError("Failed to parse CU header in DWP file"), errs());
+            joinErrors(createError("Failed to parse CU header in DWP file"),
+                       std::move(ExtractionErr)),
+            errs());
```

Comment:
Presumably this improves the error handling by producing a single error with the outer/contextual message, then the underlying root cause? (rather than printing two separate errors - inner then outer) - could that behavior/improvement be tested?

Reply:
Yes, the errors will be grouped together in this case. As you said, we would previously report the inner error through the DWARFContext and then the caller would report an error "Failed to parse CU header in DWP file". This does slightly change behavior such that "Failed to parse CU header in DWP file" will get reported first with the inner error message following afterwards.

I'll see if there are any existing tests that are suitable to extend or copy. If not, I may need help producing one. 😅

Reply:
@dwblaikie I hacked up `DWARFContext::fixupIndex` by adding an early return before any work is done and ran the llvm test suite. It doesn't look like any tests failed after that though. I'm not sure how this codepath is triggered. Do you have any suggestions here? I'd like to continue making progress improving the error handling in the LLVM DWARF Parser and refactoring LLDB to use it.

Reply:
ah, `fixupIndex` should be reachable (simplest way to find it would be to `assert(false);` in there and you'd find which tests reach the code) via tests using `-manaully-generate-unit-index` - though they might not be testing invalid headers, and such - and could be improved to do so, that'd involve some manually crafted DWARF (probably written in assembly and assembled with `llvm-mc` in the test) that has a corrupted header.

It's a bit of a chore to create a hand-crafted dwp file, though...

Reply:
Thanks for the tip! Added a test. 🙂

Reply:
I think you don't want to use `joinErrors` here. This will result in two separate errors that are reported individually, rather than one error with an outer and an inner context. A better way would be to create a new `Error` that concatenates the necessary messages. We use this trick in llvm-readobj, for example. See https://github.com/llvm/llvm-project/blob/bfcd05317d0fbe90474eda13a4dbf33c2cee4130/llvm/tools/llvm-readobj/ELFDumper.cpp#L486.

Reply:
Thanks for the pointer! I'll update this to use that trick.